design/trait-separated-from-underlying-workload-should-patch-back.md

```diff
@@ -0,0 +1,292 @@
+  template:
+    spec:
+      securityContext:
+        runAsNonRoot: true
+      nodeSelector:
+        disktype: ssd
+      containers:
+        - name: wordpress
+          image: wordpress:4.6.1-apache
+    ...
+```
+
+However, any changes happened to pod template of Deployment will trigger re-create of pods. 
+In our current workflow, traits will take effect asynchronously after workload created. That means
+a deployment will be created first without these traits' information, and then being updated several times.
+Updating pod template of deployment leads to re-create of pods, which finally makes the Application unstable. 
+
+Let's use a concrete example to make this problem more clear. For example, we currently have an OAM app running
+like below. It just contains a component with ContainerizedWorkload, and a canary trait.
+ 
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+  name: example-component
+spec:
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    spec:
+      containers:
+        - name: wordpress
+          image: wordpress:4.6.1-apache
+          ports:
+            - containerPort: 80
+              name: wordpress
+          env:
+            - name: TEST_ENV
+              value: test
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  - componentName: example-component
+      traits:
+        - trait:
+            apiVersion: extended.oam.dev/v1alpha2
+            kind: Canary
+            spec:
+              canaryNumber: 1
+```
+
+The real running workload is deployment.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+spec:
+  template:
+    spec:
+      containers:
+        - name: wordpress
+          image: wordpress:4.6.1-apache
+          ports:
+            - containerPort: 80
+              name: wordpress
+          env:
+            - name: TEST_ENV
+              value: test
+```
+
+After some time, the App operator want to add two traits on this App for some security reason, and they are both fields
+of underlying Deployment.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: example-component
+      traits:
+        - trait:
+            apiVersion: extended.oam.dev/v1alpha2
+            kind: Canary
+            spec:
+              canaryNumber: 1
+        - trait:
+            apiVersion: extended.oam.dev/v1alpha2
+            kind: NodeSelector
+            spec:
+              disktype: ssd
+        - trait:
+            apiVersion: extended.oam.dev/v1alpha2
+            kind: SecurityContext
+            spec:
+              runAsNonRoot: true
+```
+
+After AppConfig deployed, NodeSelector and SecurityContext trait will work in a random order.
+Assuming the controller of NodeSelector trait will work first, then it will patch and update the underlying Deployment.
```

Comment:
```suggestion
After the AppConfig is deployed, NodeSelector and SecurityContext trait will work in random order.
Assuming the controller of NodeSelector trait works first, then it will patch and update the underlying Deployment.
```